src/super_gradients/training/utils/detection_utils.py

```diff
@@ -717,82 +740,378 @@ class NMS_Type(str, Enum):
+    :param recall_thresholds:  Tensor of shape (max_n_rec_thresh) list of recall thresholds used to compute MaP
+    :param score_threshold:    Minimum confidence score to consider a prediction for the computation of
+                                    precision, recall and f1 (not MaP)
+    :param device:             Device
+
+    :return:
+        :precision, recall, ap: Tensors of shape (nb_iou_thrs)
+    """
+    nb_iou_thrs = preds_matched.shape[-1]
+
+    tps = preds_matched
+    fps = torch.logical_and(torch.logical_not(preds_matched), torch.logical_not(preds_to_ignore))
+
+    if len(tps) == 0:
+        return 0, 0, torch.zeros(nb_iou_thrs, device=device)
+
+    # Sort by decreasing score
+    dtype = torch.uint8 if preds_scores.is_cuda and preds_scores.dtype is torch.bool else preds_scores.dtype
+    sort_ind = torch.argsort(preds_scores.to(dtype), descending=True)
+    tps = tps[sort_ind, :]
+    fps = fps[sort_ind, :]
+    preds_scores = preds_scores[sort_ind]
+
+    # Rolling sum over the predictions
+    rolling_tps = torch.cumsum(tps, axis=0, dtype=torch.float)
+    rolling_fps = torch.cumsum(fps, axis=0, dtype=torch.float)
+
+    rolling_recalls = rolling_tps / n_targets
+    rolling_precisions = rolling_tps / (rolling_tps + rolling_fps + torch.finfo(torch.float64).eps)
+
+    # Reversed cummax to only have decreasing values
+    rolling_precisions = rolling_precisions.flip(0).cummax(0).values.flip(0)
+
+    # shape = (nb_iou_thrs, n_recall_thresholds)
+    recall_thresholds = recall_thresholds.view(1, -1).repeat(nb_iou_thrs, 1)
+
+    # We want the index i so that: rolling_recalls[i-1] < recall_thresholds[k] <= rolling_recalls[i]
+    # Note:  when recall_thresholds[k] > max(rolling_recalls), i = len(rolling_recalls)
+    # Note2: we work with transpose (.T) to apply torch.searchsorted on first dim instead of the last one
+    recall_threshold_idx = torch.searchsorted(rolling_recalls.T, recall_thresholds, right=False).T
+
+    # When recall_thresholds[k] > max(rolling_recalls), rolling_precisions[i] is not defined, and we want precision = 0
+    rolling_precisions = torch.cat((rolling_precisions, torch.zeros(1, nb_iou_thrs, device=device)), dim=0)
+
+    # shape = (n_recall_thresholds, nb_iou_thrs)
+    sampled_precision_points = torch.gather(rolling_precisions, index=recall_threshold_idx, dim=0)
+
+    # Average over the recall_thresholds
+    ap = sampled_precision_points.mean(0)
+
+    # We want the rolling precision/recall at index i so that: preds_scores[i-1] < score_threshold <= preds_scores[i]
+    # Note: torch.searchsorted works on increasing sequence and preds_scores is decreasing, so we work with "-"
+    smallest_score_idx_above_thresh = torch.searchsorted(-preds_scores, -score_threshold, right=False)
+    recall = rolling_recalls[smallest_score_idx_above_thresh]
+    precision = rolling_precisions[smallest_score_idx_above_thresh]
+
+    return precision, recall, ap
```

Comment:
A small pet peeve of mine, but can we output ap first and give it the respect it deserves?

Reply:
changed in bba66d102b960a772181b4da04ada6674b57bd48